### DISCUSSION TITLE
Compile time unrolled `@parameter for` loop?

### DISCUSSION THREAD

**Discussion by paugier** [modular/modular#1422] [Ideas]
It's possible to write

```python
    check_bench_1_nelts[2](original_time)
    check_bench_1_nelts[4](original_time)
    check_bench_1_nelts[8](original_time)
    check_bench_1_nelts[16](original_time)
    check_bench_1_nelts[32](original_time)
```

Given that there are `@parameter if` and `@parameter fn`, wouldn't it make sense to be able to write something like:

```python
@parameter
for nelts in [2, 4, 8, 16, 32]:
    check_bench_1_nelts[nelts](original_time)
```

**Comment by lattner** [comment]
I think this is supported if your for loop is unrolled but I'm not entirely sure the phase ordering here: does loop unroll happen before elaboration @weiweichen ?

**Reply by weiweichen** [reply]
loop unrolling currently happens after elaboration because we don't know the trip count till elaboration happens. Currently we only support `for` with `range`, not `for i in [...]` yet, and to get the loop bounds and step, we need elaboration to get those values out of `range`. I think a `@parameter for` makes sense if we can get the loop info before elaboration, maybe when we have support for 
```
for nelts in [2, 4, 8, 16, 32]:
```
or we can add something like `@parameter for`.

**Comment by mikowals** [comment]
There is now `VariadicList.__iter__()` to allow `for width in VariadicList[Int](2, 4, 8, 16, 32)` but `width` will be a dynamic value.  All the values are available at compile time, it is just that the workaround using `unroll` is still necessary:

```
alias simd_widths = VariadicList[Int](2, 4, 8, 16, 32)
@parameter
fn simd_loop[width_index: Int]():
    alias width = simd_widths[width_index] # VariadicList.__getitem__() produces a compile time value
    check_bench_1_nelts[width](original_time)

# VariadicList.__len__() produces a compile time value
unroll[len(simd_widths), simd_loop]()
```

Should there be a feature request to either implement `@parameter` on for loops or to make `for width in VariadicList[Int](2, 4, 8, 16, 32)` produce a compile time value without needing a decorator?

**Reply by weiweichen** [reply]
I think `@parameter` will still be needed like it is for `@parameter if` for the compiler to know that the loop iteration is a constant value, regardless whether `VariadicList.__iter__()` is a constant value or not. In other words, IMO, the key feature is to add `@parameter for` where the loop iterator can be used as a compile time constant/"parameter" in the loop body, and then `VariadicList` can leverage this new feature.
